### ISSUE TITLE
TSQL: unparsable special characters in variable declaration

### ISSUE THREAD

**Issue by dsbudiac** [sqlfluff/sqlfluff#4840]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

TSQL allows for some special characters like `@`, `#`, `$` in variable declarations i.e.

```sql
DECLARE @$variable int = 1
```
More info:
https://stackoverflow.com/questions/37896780/how-to-use-special-characters-in-sql-server-variables

sqlfluff throws an "Found unparsable section" error.

Personally, I use these special characters to denote variable parameters that have been coerced.

### Expected Behaviour

Allow for some special characters in TSQL.

### Observed Behaviour

sqlfluff throws an "Found unparsable section" error.

### How to reproduce

Set dialect to `tsql` and try to lint a file like:

```sql
DECLARE @$var1 int
DECLARE @@var2 int
DECLARE @#var3 int
```

### Dialect

TSQL

### Version

2.0.7

### Configuration

[sqlfluff]
dialect = tsql
exclude_rules = LT01, LT02, RF02, LT04, LT08
warnings = RF06, ST05, ST03

[sqlfluff:rules:references.special_chars]
additional_allowed_characters = .

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
